src/Libraries/RevitNodes/Elements/DirectShape.cs

```diff
@@ -0,0 +1,403 @@
+        /// </summary>
+        /// <param name="shape"></param>
+        private void InteralSetDirectShape(Autodesk.Revit.DB.DirectShape shape)
+        {
+            this.InternalDirectShape = shape;
+            this.InternalElementId = shape.Id;
+            this.InternalUniqueId = shape.UniqueId;
+        }
+
+        /// <summary>
+        /// Sets the internalDirectShape to point to some geometry,
+        /// this method also generates tessellated geometry from the protogeometry object
+        /// and sets the material of the generated Revit faces
+        /// </summary>
+        /// <param name="shapeReference"></param>
+        /// <param name="materialId"></param>
+        private void InteralSetShape(DesignScriptEntity shapeReference, ElementId materialId)
+        {
+            //if the elementID for the current directShape revitElement exists on the input Geometry AND
+            //the value stored at that key is equal to the materialId we're trying to set then we don't
+            //need to regenerate the geometry
+            if((shapeReference.Tags.LookupTag(this.InternalElementId.ToString()) as ElementId) == materialId)
+            {
+                return;
+            }
+
+            TransactionManager.Instance.EnsureInTransaction(Document);
+
+            IList<GeometryObject> tessellatedShape = GenerateTessellatedGeo(shapeReference, materialId);
+
+            this.InternalDirectShape.SetShape(tessellatedShape);
+            shapeReference.Tags.AddTag(this.InternalElementId.ToString(), materialId);
+            TransactionManager.Instance.TransactionTaskDone();
+        }
+
+        /// <summary>
+        /// Sets the internalDirectShape to have a new name
+        /// </summary>
+        /// <param name="name"></param>
+        private void InteralSetName(string name)
+        {
+          
+            if (name != this.InternalDirectShape.Name)
+            {
+                TransactionManager.Instance.EnsureInTransaction(Document);
+                this.InternalDirectShape.SetName(name);
+                TransactionManager.Instance.TransactionTaskDone();
+            }
+
+            
+        }
+        #endregion
+
+        #region Public properties
+
+        /// <summary>
+        /// Gets the centroid of the geometry contained in DirectShape 
+        /// </summary>
+        public Point Centroid
+        {
+            get
+            {
+                TransactionManager.Instance.EnsureInTransaction(DocumentManager.Instance.CurrentDBDocument);
+                DocumentManager.Regenerate();
+                var Revitbb = this.InternalElement.get_BoundingBox(null);
+                TransactionManager.Instance.TransactionTaskDone();
+                var bb = Revitbb.ToProtoType();
+                var cube = bb.ToCuboid();
+                var point = cube.Centroid();
+                bb.Dispose();
+                cube.Dispose(); 
+              
+                return point;
```

Comment:
This would not actually calculate the centroid, simply the bounding box centroid. I don't see a strong reason to have this property anyway, let's remove it.